File: backend/src/providers/finnhub.ts
import { Ticker, NewsArticle, StockBasicFinancials } from "@shared/types";

const BASE_URL = "https://finnhub.io/api/v1";

type FinnhubNewsCategory =
  | "general"
  | "forex"
  | "crypto"
  | "merger"
  | "technology"
  | "business"
  | "top news";

interface FinnhubNewsArticle {
  category: FinnhubNewsCategory;
  datetime: number;
  headline: string;
  id: number;
  image: string;
  related: unknown;
  source: string;
  summary: string;
  url: string;
}

interface FinnhubMetricResponse {
  symbol: string;
  metric: {
    "10DayAverageTradingVolume"?: number;
    "52WeekHigh"?: number;
    "52WeekLow"?: number;
    beta?: number;
    bookValuePerShareAnnual?: number;
    cashFlowPerShareAnnual?: number;
    currentDividendYieldTTM?: number;
    currentRatioAnnual?: number;
    currentRatioQuarterly?: number;
    dividendPerShareAnnual?: number;
    dividendYieldIndicatedAnnual?: number;
    enterpriseValue?: number;
    epsGrowthTTMYoy?: number;
    epsTTM?: number;
    evEbitdaTTM?: number;
    evRevenueTTM?: number;
    forwardPE?: number;
    forwardPEG?: number;
    grossMarginAnnual?: number;
    "longTermDebt/equityAnnual"?: number;
    marketCapitalization?: number;
    netInterestCoverageAnnual?: number;
    netProfitMarginAnnual?: number;
    operatingMarginAnnual?: number;
    payoutRatioAnnual?: number;
    pb?: number;
    peTTM?: number;
    pegTTM?: number;
    psTTM?: number;
    quickRatioAnnual?: number;
    quickRatioQuarterly?: number;
    revenueGrowthTTMYoy?: number;
    revenuePerShareAnnual?: number;
    roeRfy?: number;
    "totalDebt/totalEquityAnnual"?: number;
  };
}

function transformMetrics(m: FinnhubMetricResponse["metric"]): StockBasicFinancials {
  return {
    peTTM: m.peTTM,
    forwardPE: m.forwardPE,
    pb: m.pb,
    priceToSalesTTM: m.psTTM,
    pegTTM: m.pegTTM,
    evToEBITDA: m.evEbitdaTTM,
    evToRevenue: m.evRevenueTTM,
    marketCap: m.marketCapitalization,
    enterpriseValue: m.enterpriseValue,
    epsTTM: m.epsTTM,
    roe: m.roeRfy,
    netProfitMargin: m.netProfitMarginAnnual,
    grossMargin: m.grossMarginAnnual,
    operatingMargin: m.operatingMarginAnnual,
    revenueGrowth: m.revenueGrowthTTMYoy,
    epsGrowth: m.epsGrowthTTMYoy,
    debtToEquity: m["totalDebt/totalEquityAnnual"],
    currentRatio: m.currentRatioAnnual,
    quickRatio: m.quickRatioAnnual,
    interestCoverage: m.netInterestCoverageAnnual,
    bookValue: m.bookValuePerShareAnnual,
    revenuePerShare: m.revenuePerShareAnnual,
    dividendPerShare: m.dividendPerShareAnnual,
    divYieldTTM: m.currentDividendYieldTTM,
    payoutRatio: m.payoutRatioAnnual,
    yearHigh: m["52WeekHigh"],
    yearLow: m["52WeekLow"],
    adtv10Day: m["10DayAverageTradingVolume"],
    beta: m.beta,
  };
}

export async function getMarketNews(
  category: FinnhubNewsCategory,
): Promise<NewsArticle[]> {
  const url = new URL(`${BASE_URL}/news`);
  url.searchParams.set("category", category);
  url.searchParams.set("token", process.env.FINNHUB_API_KEY!);

  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`HTTP ${res.status}: Failed to fetch market news.`);
  }

  const data = (await res.json()) as FinnhubNewsArticle[];
  return data.map((article) => ({
    ...article,
    datetime: new Date(article.datetime * 1000),
  }));
}

export async function getBasicFinancials(
  ticker: Ticker,
): Promise<StockBasicFinancials | undefined> {
  const url = new URL(`${BASE_URL}/stock/metric`);
  url.searchParams.set("token", process.env.FINNHUB_API_KEY!);
  url.searchParams.set("symbol", ticker.symbol);
  url.searchParams.set("metric", "all");

  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`HTTP ${res.status}: Failed to fetch financials for ${ticker.symbol}.`);
  }

  const data = (await res.json()) as FinnhubMetricResponse;
  return transformMetrics(data.metric);
}

File: backend/app/providers/finnhub.py
import asyncio
import os
from datetime import datetime
from typing import Any

import finnhub

from app.schemas.news import NewsArticle, NewsCategory
from app.schemas.stock import StockBasicFinancials, Ticker


def _client() -> finnhub.Client:
    api_key = os.environ.get("FINNHUB_API_KEY")
    if not api_key:
        raise RuntimeError("Finnhub API key is undefined.")
    return finnhub.Client(api_key=api_key)


def _transform_metrics(metric: dict[str, Any]) -> StockBasicFinancials:
    return StockBasicFinancials(
        peTTM=metric.get("peTTM"),
        forwardPE=metric.get("forwardPE"),
        pb=metric.get("pb"),
        priceToSalesTTM=metric.get("psTTM"),
        pegTTM=metric.get("pegTTM"),
        evToEBITDA=metric.get("evEbitdaTTM"),
        evToRevenue=metric.get("evRevenueTTM"),
        marketCap=metric.get("marketCapitalization"),
        enterpriseValue=metric.get("enterpriseValue"),
        epsTTM=metric.get("epsTTM"),
        roe=metric.get("roeRfy"),
        netProfitMargin=metric.get("netProfitMarginAnnual"),
        grossMargin=metric.get("grossMarginAnnual"),
        operatingMargin=metric.get("operatingMarginAnnual"),
        revenueGrowth=metric.get("revenueGrowthTTMYoy"),
        epsGrowth=metric.get("epsGrowthTTMYoy"),
        debtToEquity=metric.get("totalDebt/totalEquityAnnual"),
        currentRatio=metric.get("currentRatioAnnual"),
        quickRatio=metric.get("quickRatioAnnual"),
        interestCoverage=metric.get("netInterestCoverageAnnual"),
        bookValue=metric.get("bookValuePerShareAnnual"),
        revenuePerShare=metric.get("revenuePerShareAnnual"),
        dividendPerShare=metric.get("dividendPerShareAnnual"),
        divYieldTTM=metric.get("currentDividendYieldTTM"),
        payoutRatio=metric.get("payoutRatioAnnual"),
        yearHigh=metric.get("52WeekHigh"),
        yearLow=metric.get("52WeekLow"),
        adtv10Day=metric.get("10DayAverageTradingVolume"),
        beta=metric.get("beta"),
    )


async def get_market_news(category: NewsCategory) -> list[NewsArticle]:
    client = _client()
    articles = await asyncio.to_thread(client.general_news, category, min_id=0)
    return [
        NewsArticle(
            id=article["id"],
            headline=article["headline"],
            datetime=datetime.fromtimestamp(article["datetime"]),
            url=article["url"],
            source=article["source"],
            category=article.get("category"),
            image=article.get("image"),
            summary=article.get("summary"),
        )
        for article in articles
    ]


async def get_basic_financials(ticker: Ticker) -> StockBasicFinancials | None:
    client = _client()
    data = await asyncio.to_thread(
        client.company_basic_financials, ticker.symbol, "all"
    )
    metric = data.get("metric")
    if not metric:
        return None
    return _transform_metrics(metric)
